src/core/server/http/lifecycle_handlers.test.ts

```diff
@@ -0,0 +1,275 @@
+      const config = createConfig({ xsrf: { whitelist: [], disableProtection: false } });
+      const handler = createXsrfPostAuthHandler(config);
+      const request = forgeRequest({ method: 'post', headers: { 'kbn-version': 'some-version' } });
+
+      toolkit.next.mockReturnValue('next' as any);
+
+      const result = handler(request, responseFactory, toolkit);
+
+      expect(responseFactory.badRequest).not.toHaveBeenCalled();
+      expect(toolkit.next).toHaveBeenCalledTimes(1);
+      expect(result).toEqual('next');
+    });
+
+    it('returns a bad request if called without xsrf or version header', () => {
+      const config = createConfig({ xsrf: { whitelist: [], disableProtection: false } });
+      const handler = createXsrfPostAuthHandler(config);
+      const request = forgeRequest({ method: 'post' });
+
+      responseFactory.badRequest.mockReturnValue('badRequest' as any);
+
+      const result = handler(request, responseFactory, toolkit);
+
+      expect(toolkit.next).not.toHaveBeenCalled();
+      expect(responseFactory.badRequest).toHaveBeenCalledTimes(1);
+      expect(responseFactory.badRequest.mock.calls[0][0]).toMatchInlineSnapshot(`
+        Object {
+          "body": "Request must contain a kbn-xsrf header.",
+        }
+      `);
+      expect(result).toEqual('badRequest');
+    });
+
+    it('accepts requests if protection is disabled', () => {
+      const config = createConfig({ xsrf: { whitelist: [], disableProtection: true } });
+      const handler = createXsrfPostAuthHandler(config);
+      const request = forgeRequest({ method: 'post', headers: {} });
+
+      toolkit.next.mockReturnValue('next' as any);
+
+      const result = handler(request, responseFactory, toolkit);
+
+      expect(responseFactory.badRequest).not.toHaveBeenCalled();
+      expect(toolkit.next).toHaveBeenCalledTimes(1);
+      expect(result).toEqual('next');
+    });
+
+    it('accepts requests if path is whitelisted', () => {
+      const config = createConfig({
+        xsrf: { whitelist: ['/some-path'], disableProtection: false },
+      });
+      const handler = createXsrfPostAuthHandler(config);
+      const request = forgeRequest({ method: 'post', headers: {}, path: '/some-path' });
+
+      toolkit.next.mockReturnValue('next' as any);
+
+      const result = handler(request, responseFactory, toolkit);
+
+      expect(responseFactory.badRequest).not.toHaveBeenCalled();
+      expect(toolkit.next).toHaveBeenCalledTimes(1);
+      expect(result).toEqual('next');
+    });
+  });
+});
+
+describe('versionCheck post-auth handler', () => {
+  let toolkit: ReturnType<typeof httpServerMock.createToolkit>;
+  let responseFactory: ReturnType<typeof httpServerMock.createLifecycleResponseFactory>;
+
+  beforeEach(() => {
+    toolkit = httpServerMock.createToolkit();
+    responseFactory = httpServerMock.createLifecycleResponseFactory();
+  });
+
+  it('forward the request to the next interceptor if header matches', () => {
```

Comment:
I suspect this functionality is covered by the integration tests. Isn't it? If yes, I'd remove the unit tests because it's hard to write them due to coupling with hapi.

Reply:
Yes, it's tested in integration tests. However I don't think we are really tightly coupled with hapi here, as all used interfaces/types are from core (even if we mimic hapi's types for some of them). I'm leaning to keep them as this is the only place we can really test them unitarily (integ test have all interceptors registered)